spark/src/main/java/org/apache/iceberg/spark/data/SparkValueReaders.java

```diff
@@ -334,22 +351,20 @@ public InternalRow read(Decoder decoder, Object reuse) throws IOException {
       }
 
       // otherwise, we need to return an InternalRow as a struct data
-      InternalRow struct = new GenericInternalRow(nullIndex >= 0 ? alts.size() : alts.size() + 1);
+      int numOfFields = isTagFieldProjected ?
+          projectedFieldIdsToIdxInReturnedRow.size() + 1 : projectedFieldIdsToIdxInReturnedRow.size();
+      InternalRow struct = new GenericInternalRow(numOfFields);
       for (int i = 0; i < struct.numFields(); i += 1) {
         struct.setNullAt(i);
       }
+      int fieldId = (nullIndex < 0 || index < nullIndex) ? index : index - 1;
+      if (isTagFieldProjected) {
+        struct.setInt(0, fieldId);
+      }
 
       Object value = readers[index].read(decoder, reuse);
-
-      if (nullIndex < 0) {
-        struct.update(index + 1, value);
-        struct.setInt(0, index);
-      } else if (index < nullIndex) {
-        struct.update(index + 1, value);
-        struct.setInt(0, index);
-      } else {
-        struct.update(index, value);
-        struct.setInt(0, index - 1);
+      if (projectedFieldIdsToIdxInReturnedRow.containsKey(fieldId)) {
+        struct.update(projectedFieldIdsToIdxInReturnedRow.get(fieldId), value);
```

Comment:
I am not sure if this line would impact performance since it uses a hashmap lookup. Maybe another reason to use an array.

Reply:
Agree.